File: src/main/java/ru/nsu/fit/g20204/kuznetsov/wireframe/b_spline_editor/BSplineEditor.java
package ru.nsu.fit.g20204.kuznetsov.wireframe.b_spline_editor;

import ru.nsu.fit.g20204.kuznetsov.wireframe.math.BSpline;
import ru.nsu.fit.g20204.kuznetsov.wireframe.model.Geometry;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class BSplineEditor extends JFrame {
    private final BSplinePane splinePane = new BSplinePane();
    private final BSplineParamPane paramPane = new BSplineParamPane(splinePane, this);
    private final List<Function<Geometry, Void>> splineChangedListeners = new ArrayList<>();


    public BSplineEditor() {
        super("B-Spline Editor");

        this.setMinimumSize(new Dimension(640, 480));
        this.setLocation(600, 160);
        this.setLocation(600, 160);
        this.setVisible(true);

        this.add(splinePane, BorderLayout.CENTER);
        this.add(paramPane, BorderLayout.PAGE_END);
    }

    public BSpline getSpline() {
        return splinePane.getSpline();
    }

    public Geometry getSplineModel() {
        return paramPane.getSplineModel();
    }

    public void addSplineModelChangeListener(Function<Geometry, Void> listener) {
        splineChangedListeners.add(listener);
    }

    public void applySpline() {
        for (var l : splineChangedListeners) {
            l.apply(paramPane.getSplineModel());
        }
    }
}

File: src/main/java/ru/nsu/fit/g20204/kuznetsov/wireframe/b_spline_editor/BSplineEditor.kt
package ru.nsu.fit.g20204.kuznetsov.wireframe.b_spline_editor

import ru.nsu.fit.g20204.kuznetsov.wireframe.math.BSpline
import ru.nsu.fit.g20204.kuznetsov.wireframe.model.Geometry
import java.awt.BorderLayout
import java.awt.Dimension
import javax.swing.JFrame

class BSplineEditor : JFrame("B-Spline Editor") {
    private val splinePane = BSplinePane()
    private val paramPane = BSplineParamPane(splinePane, this)
    private val splineChangedListeners: MutableList<(Geometry) -> Unit> = ArrayList()

    init {
        minimumSize = Dimension(640, 480)
        setLocation(600, 160)
        setLocation(600, 160)
        isVisible = true
        add(splinePane, BorderLayout.CENTER)
        add(paramPane, BorderLayout.PAGE_END)
    }

    val spline: BSpline
        get() = splinePane.spline
    val splineModel: Geometry
        get() = paramPane.splineModel

    fun addSplineModelChangeListener(listener: (Geometry) -> Unit) {
        splineChangedListeners.add(listener)
    }

    fun applySpline() {
        for (l in splineChangedListeners) {
            l(paramPane.splineModel)
        }
    }
}
